proposals/user-scripts-execute-api.md

```diff
@@ -0,0 +1,150 @@
+  frameId: number,
+  // The result of the script execution.
+  result?: any,
+  // The error, if any. error and result are mutually exclusive.
+  error?: any,
+}
+```
+
+#### Methods
+
+```
+// Injects a script into a target context. The script will be run at
+// `document_idle` unless the `injectImmediately` property is set. If the
+// script evaluates to a promise, the browser will wait
+// for the promise to settle and return the resulting value.
+Promise<InjectionResult[]> browser.userScripts.execute(
+  // The details of the script which to inject
+  injection: UserScriptInjection
+)
+```
+
+Note: Promises are supported in Manifest V3 and later, but callbacks are provided for backward compatibility. You cannot use both on the same function call. The promise resolves with the same type that is passed to the callback.
+
+### New Permissions
+
+No new permissions will be added; this will be part of the existing `userScripts` permission
+
+### Manifest Changes
+
+N/A
+
+### Custom Bindings
+
+N/A
+
+## User Experience
+
+### UI Elements and User-Visible Effects
+
+There are no new UI elements or user-visible effects (exempting those caused by the script of the extension).
+
+## Security and Privacy
+
+### Exposed Sensitive Data
+
+This API does not directly expose any new sensitive data, as it relies on the host permissions of the extension and the userScripts permission. If the extension already has host permissions to a site, it can access all the data associated with that site (including scraping the DOM, accessing cookies, modifying network requests, and more).
+
+### Abuse Scenarios
+
+With any script injection API, there are significant abuse possibilities - such as injecting malicious JavaScript intended to collect user data maliciously.  Additionally, this allows extensions to inject code that's not contained within the extension's package. However, this is inline with the existing userScripts API, and there are no new abuse mechanisms introduced as a part of this API.
+
+#### Mitigations
+The existing mitigations for host permissions - most dominantly, runtime host permissions - apply here.
+
+### Additional Security Considerations
+
+N/A
+
+## Alternatives
+
+### Past Workarounds
+
+In the past, developers leveraged `chrome.tabs.executeScript()`. However, this method was removed in MV3.
+
+### Existing Workarounds
+
+Developers can utilize `userScripts.register()` to inject JavaScript into a known set of hosts, but this doesn't cover the case of programmatic injections to a specific target or one-time injections.
+
+### Open Web API
+
+Since this API exposes the ability to inject JavaScript on sites that are not in the same origin, it does not belong on the web.
+
+## (Potential) Future Enhacements
+
+### Add func to `ScriptSource`
+
+Add `func` and `args` property to `ScriptSource` to specify a JavaScript function to inject. This could be used both by `userScripts.register()` and `userScripts.execute()`.
```

Comment:
It's worth highlighting that this isn't as useful in this context, since the extension can trivially construct a string that enables this same functionality.  Additionally, if the extension is using func + args, it isn't using remotely-hosted code, and could also use `scripting.executeScript()`.  The only case it would need to use this in that scenario would be to inject in the user script world.

Reply:
...also when Chrome makes `args` use the structured clone algorithm.

Reply:
Yep, that's a reasonable point, too